Uitid/pk 41 keycloak switch (#408)
cultuurnet/apidocs: projects/authentication/docs/user-access-token.md

> Additionally, we will need to configure the following settings for your client on our end:

Reviewer: ```suggestion
Additionally, you will need to configure the following settings for your client on [Publiq Platform](https://platform.publiq.be/). Select your integration and select the "settings" tab.
```

Reviewer 2: I think this is partially context dependent. Not familiar enough with the UDB context here, but for UiTPAS, the user permissions will have to be set up in the UiTPAS Admin. At this point, this isn't possible on publiq platform.

Are user permissions even a thing in UDB? This might be related to roles (and that would make sense to align it, so UPS and UDB have a similar logic). But this is out of scope of this request.

Reviewer: User permissions are managed directly in UDB, specifically there is a list of emails that are the admins.
The text below is referring to login, callback and logout urls, and those are managable in Publiq Platform.
Maybe we should just move this outside this PR and create a seperate ticket for this, it is not directly related to Keycloak.

Author: > Maybe we should just move this outside this PR and create a seperate ticket for this, it is not directly related to Keycloak.

yes please 🙏 
I don't think I changed that line :)

Reviewer: I made a ticket https://jira.publiq.be/browse/PPF-638 so ignore this remark for now.

* **Login URL**: In some cases the authorization server will need to redirect the user back to a login URL on your application. This URL should point to a route in your application that ends up redirecting to the `/realms/uitid/protocol/openid-connect/auth` endpoint on publiq's authorization server, e.g. `https://example.com/login`. Note that it requires `https` and it cannot point to `localhost`. It can include query parameters and a URI fragment.
* **Callback URL(s)**: The absolute URL(s) of the page(s) where your users can be redirected back to after they log in. You can specify any callback URL whenever you redirect a user to the authorization server to log in, but it needs to be registered on our end first to prevent phishing attacks. For example `https://example.com/authorize` for your production application, and/or `https://localhost:8080/authorize` for local development.

> * **Logout URL(s)**: The absolute URL(s) of the page(s) where you users can be redirected back to after they log out. You can specify any logout URL whenever you redirect a user to the authorization server to log out, but it needs to be registered on our end first to prevent phishing attacks. For example `https://example.com` if you want to redirect the user back to the homepage of your application after logging out, or `https://example.com/logged-out` if you want to redirect them to a page with a confirmation message like "You're logged out!" after logging out. Like callback URL(s), logout URL(s) may use `https://localhost` for local development.

Reviewer: Maybe add a line that it is end-slash sensitive ?
Common error.

Author: done

    autonumber
    actor User
    User-->>Client: Click login link

>     Client->>Auth server: Redirect to GET /auth

Reviewer: ```suggestion
    Client->>Auth server: Redirect to GET /realms/uitid/protocol/openid-connect/auth
```

    Auth server-->>User: Show login page
    User-->>Auth server: Login and give consent
    Auth server-->>Client: Redirect back to callback URL with authorization code

>     Client->>Auth server: POST /token with authorization code<br /> and client credentials

Reviewer: ```suggestion
    Client->>Auth server: POST /realms/uitid/protocol/openid-connect/token with authorization code<br /> and client credentials
```

Author: This is the mermaid diagram and I explicitly chose to leave out the full path. Maybe I can reduce confusion further by getting rid of the `/` 

This is how it looks like with the full path:
<img width="590" alt="image" src="https://github.com/user-attachments/assets/013648b7-2bbc-4c98-bd96-199120f04314">

Author: Changed to full path

#### More info

publiq uses [Keycloak](https://www.keycloak.org/) as its authentication and authorization server, which implements the [OpenID Connect specification](https://openid.net/specs/openid-connect-core-1_0.html). To learn more about the Authorization Code Flow, see the [Authentication using the Authorization Code Flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth) section of the OpenID Connect specification.

<!-- theme: success -->

> ##### SDK
>

> > It is a good practice to use an OpenID Connect compliant client SDK. SDKs exist in different languages and flavors, so the best choice greatly depends on your specific application. The OpenID.net website contains a list of [certicied](https://openid.net/developers/certified-openid-connect-implementations/) and [uncertified](https://openid.net/developers/uncertified-openid-connect-implementations/) SDKs. To configure your SDK you might need this configuration: `https://account-test.uitid.be/.well-known/openid-configuration`.

Reviewer: Same remark regarding mentioning Keycloak.

    actor User
    User-->>Client: Click login link
    Client->>Client: Generate code_verifier and code_challenge

>     Client->>Auth server: Redirect to GET /auth with code_challenge

Reviewer: ```suggestion
    Client->>Auth server: Redirect to GET /realms/uitid/protocol/openid-connect/auth with code_challenge
```

    Auth server-->>User: Show login page
    User-->>Auth server: Login and give consent
    Auth server-->>Client: Redirect back to callback URL with authorization code

>     Client->>Auth server: POST /token with authorization code,<br /> client id and code_verifier

Reviewer: ```suggestion
    Client->>Auth server: POST /realms/uitid/protocol/openid-connect/token with authorization code,<br /> client id and code_verifier
```

#### More info

> publiq uses [Keycloak](https://www.keycloak.org/) as its authentication and authorization server. To learn more about the Authorization Code Flow with PKCE, see [RFC 7636: Proof Key for Code Exchange](https://oauth.net/2/pkce/).

Reviewer: Publiq with a capital letter, and same remark regarding Keycloak.

Author: is `publiq` writtern with a capital P? :) (it was already there in small caps, and I see regularly see it small caps even on publiq.be)

When the user of your application wants to log out, clear any session data in your application including the user's access token and refresh token. If you have cached the [user's info](#user-info), make sure to also clear that.

Afterward, you should **redirect** the user to the `/realms/uitid/protocol/openid-connect/logout` URL on the authorization server so the user's session is also cleared there. This redirect will be invisible to the end user, as the authorization server will simply clear the user's session and then redirect back to your application based on a `?post_logout_redirect_uris=...` URL parameter that you can specify. Note that you can only use an allowed "logout URL" as value for the `post_logout_redirect_uris` URL parameter. (See [Requirements](#requirements))

> 

Reviewer: Maybe also add a paragraph about **id_token_hint** ?

Author: Fair point. It's mentioned in the link behind the link behind `OpenID Connect RP-Initiated Logout` in the following paragraph, but I can make it more explicit.

Author: done
